File: R/scrape.r
library(data.table)
library(RSelenium)
library(rvest)


# Crear un data table con los datos
dataTable <- data.table(
  Titulo = character(0),
  Autor = character(0),
  Revista = character(0),
  Año = integer(0),
  Editor = character(0),
  URL = character(0),
  Citado = integer(0)
)

# Inicializa el navegador Selenium
driver <- rsDriver(browser = "chrome")
remote_driver <- driver[["client"]]

# Navega a la página
remote_driver$navigate("https://scholar.google.com")


# Esperar hasta que los resultados de la búsqueda se carguen
Sys.sleep(120)

# Extraer todos los artículos de una página
articulos <- remote_driver$findElements("css selector", "#gs_res_ccl h3")
data <- list()

for (i in 1:length(articulos)) {
  title <- articulos[[i]]$getElementText()
  divIndex <- i
  tryCatch({
    link <- remote_driver$findElement("css selector", paste0("#gs_res_ccl_mid div:nth-child(", divIndex, ") div:nth-child(2) div:nth-child(3) a:nth-child(3)"))$getElementAttribute("href")
    cat("Link:", link, "\n")
  }, error = function(e) {
    cat("Error al encontrar el elemento:", e$message, "\n")
    link <- NA
  })
  
  xpath <- paste0("#gs_res_ccl_mid div:nth-child(", divIndex, ") div:nth-child(2) div:nth-child(3) a:nth-child(3)")
  citadoPor <- tryCatch({
    citadoPor <- articulos[i]$findElement("css selector", xpath)$getElementText()
    citadoPor
  }, error = function(e) {
    "No disponible"
  })
  
  numCitas <- as.integer(gsub("\\D", "", citadoPor))
  if (is.na(numCitas)) numCitas <- 0
  
  cat("Citado por:", numCitas, "\n")
  
  tryCatch({
    enlaceCitar <- articulos[i]$findElement("css selector", paste0("#gs_res_ccl_mid div:nth-child(", divIndex, ") div:nth-child(2) div:nth-child(3) a:nth-child(2)"))
    cat("Enlace Citar:", enlaceCitar$getElementAttribute("href"), "\n")
    Sys.sleep(sample(10:15, 1))
    enlaceCitar$click()
    
    tryCatch({
      Sys.sleep(sample(10:15, 1))
      enlaceBibTeX <- remote_driver$findElement("css selector", "#gs_citi a:nth-child(1)")
      cat("Enlace BibTeX:", enlaceBibTeX$getElementAttribute("href"), "\n")
      enlaceBibTeX$click()
      Sys.sleep(sample(10:15, 1))
      
      # Guardar los datos de la página en la variable bibtex_data
      bibtex_data <- remote_driver$getPageSource()[[1]]
      
      # Utilizamos rvest para extraer los datos BibTeX
      bibtex_data <- bibtex_data %>%
        read_html() %>%
        html_element("pre") %>%
        html_text()
      cat("Datos BibTeX:\n")
      cat(bibtex_data, "\n")
      data[[i]] <- bibtex_data
      
      tryCatch({
        # Analizar el BibTeX y acceder a los campos
        parsed_entry <- parse.bib(text = bibtex_data)
        if (length(parsed_entry) > 0) {
          entry <- parsed_entry[[1]]
          # Crear una nueva fila en el data table
          newDataRow <- data.table(
            Titulo = entry$Title,
            Autor = entry$Author,
            Revista = entry$Journal,
            Año = entry$Year,
            Editor = entry$Publisher,
            URL = link,
            Citado = numCitas
          )
          dataTable <- rbind(dataTable, newDataRow)
        } else {
          # Manejar el caso en que no se encuentren datos BibTeX válidos
          cat("No se encontraron datos BibTeX válidos para el artículo:", title, "\n")
          data[[i]] <- NA
        }
      }, error = function(e) {
        cat("Error al procesar los datos BibTeX para el artículo:", title, "\n")
        data[[i]] <- NA
      })
      
      remote_driver$back()
      Sys.sleep(sample(15:20, 1))
    }, error = function(e) {
      cat("No se encontró el enlace 'BibTeX' para el artículo:", title, "\n")
      data[[i]] <- NA
    })
    
    remote_driver$back()
    Sys.sleep(sample(20:35, 1))
  }, error = function(e) {
    cat("No se encontró el enlace 'Citar' para el artículo:", title, "\n")
  })
}

# Almacenar el data table en un archivo CSV
fwrite(dataTable, "web_scraping_citas.csv")

# Cerrar el navegador Selenium
driver$server$stop()

# Mostrar los primeros registros del data table
cat("Datos obtenidos:", head(dataTable))

File: scrape.js
const puppeteer = require('puppeteer');
const fs = require('fs');
const { Parser } = require('json2csv');

(async () => {
  const browser = await puppeteer.launch({ headless: false });
  const page = await browser.newPage();

  await page.goto('https://scholar.google.com/scholar?cites=5866269323493626547&as_sdt=2005&sciodt=0,5&hl=es');
  
  await page.waitForSelector('#gs_res_ccl', { timeout: 10000 });

  const articulos = await page.$x('//*[@id="gs_res_ccl"]//h3');

  const data = [];

  for (let i = 0; i < articulos.length; i++) {
    const title = await (await articulos[i].getProperty('textContent')).jsonValue();
    const div_index = i + 1;

    try {
      const linkElement = await page.waitForXPath(`//*[@id="gs_res_ccl_mid"]/div[${div_index}]/div[2]/div[3]/a[3]`);
      const link = await linkElement.getProperty('href');
      console.log(await link.jsonValue());
    } catch (e) {
      console.error("Error al encontrar el elemento:", e);
    }

    const citadoPorElement = await articulos[i].$x(`//*[@id="gs_res_ccl_mid"]/div[${div_index}]/div[2]/div[3]/a[3]`);
    const citadoPor = citadoPorElement.length > 0 ? await (await citadoPorElement[0].getProperty('textContent')).jsonValue() : "No disponible";

    const match = citadoPor.match(/\d+/);
    const num_citas = match ? parseInt(match[0]) : 0;

    console.log("Citado por:", num_citas);

    try {
        const enlaceCitar = await page.waitForXPath(`//*[@id="gs_res_ccl_mid"]/div[${div_index}]/div[2]/div[3]/a[2]`);
        console.log("Enlace Citar:", await enlaceCitar.getProperty('href').jsonValue());
        const secondsPause = randomInt(10, 15);
        await sleep(secondsPause);
        enlaceCitar.click();
        
      try {
        const enlaceBibTeX = await page.waitForXPath('//*[@id="gs_citi"]/a[1]');
        console.log("Enlace BibTeX:", await enlaceBibTeX.getProperty('href').jsonValue());
        await enlaceBibTeX.click();
        const secondsPause = randomInt(10, 15);
        await sleep(secondsPause);
        
        const bibtex_data = await page.evaluate(() => document.querySelector('#gs_citi pre').textContent);
        console.log("Datos BibTeX:");
        console.log(bibtex_data);

        try {
          const parsed_entry = JSON.parse(bibtex_data);
          const new_data = [{
            'Titulo': parsed_entry.title,
            'Autor': parsed_entry.author,
            'Revista': parsed_entry.journal,
            'Año': parsed_entry.year,
            'Editor': parsed_entry.publisher,
            'URL': link,
            'Citado': num_citas,
          }];

          data.push(new_data);
        } catch (e) {
          console.error("Error al procesar los datos BibTeX:", e);
          data.push(null);
        }

        await page.goBack();
        secondsPause = randomInt(15, 20);
        await sleep(secondsPause);
      } catch (e) {
        console.error("No se encontró el enlace 'BibTeX' para el artículo:", e);
        data.push(null);
      }

      await page.goBack();
      secondsPause = randomInt(20, 35);
      await sleep(secondsPause);
    } catch (e) {
      console.error("No se encontró el enlace 'Citar' para el artículo:", e);
    }
  }

  // Almacenar los datos en un archivo CSV
  const fields = ['Titulo', 'Autor', 'Revista', 'Año', 'Editor', 'URL', 'Citado'];
  const json2csv = new Parser({ fields });
  const csv = json2csv.parse(data);
  fs.writeFileSync('web_scraping_citas.csv', csv, 'utf-8');
  await sleep(60000);

  await browser.close();

  console.log("Datos obtenidos: ");
  console.log(data);
})();

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

function randomInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}
